File: game/scripts/vscripts/modifiers/upgrades/items/modifier_upgrade_item_yasha.lua
modifier_upgrade_item_yasha = class({})

function modifier_upgrade_item_yasha:RemoveOnDeath() return false end
function modifier_upgrade_item_yasha:IsPurgable() return false end

function modifier_upgrade_item_yasha:OnCreated()
    self.speed_buff_pct = 30
	self.as_speed = 80
    self.damage = 3
end

function modifier_upgrade_item_yasha:OnRefresh()
    if IsServer() then
        self:IncrementStackCount()
    end
end

function modifier_upgrade_item_yasha:DeclareFunctions()
    return { 
		MODIFIER_PROPERTY_MOVESPEED_BONUS_PERCENTAGE,
        MODIFIER_PROPERTY_ATTACKSPEED_BONUS_CONSTANT,
    }
end

function modifier_upgrade_item_yasha:GetModifierAttackSpeedBonus_Constant()
	return self.as_speed
end

function modifier_upgrade_item_yasha:GetModifierMoveSpeedBonus_Percentage(params)
    return self.speed_buff_pct
end

function modifier_upgrade_item_yasha:GetModifierPreAttack_BonusDamage(params)
    return self.damage
end

function modifier_upgrade_item_yasha:GetTexture()
    return 'item_yasha'
end

File: src/vscripts/modifiers/upgrades/items/modifier_upgrade_item_yasha.ts
import { registerModifier } from "../../../lib/dota_ts_adapter";
import { ModifierUpgrade } from "../modifier_upgrade";

@registerModifier("modifier_upgrade_item_yasha")
class ModifierItemYasha extends ModifierUpgrade {
    speedBuffPct = 30;
    asSpeed = 80;
    damage = 3;

    OnRefresh() {
        if (IsServer()) {
            this.IncrementStackCount();
        }
    }

    DeclareFunctions() {
        return [
            ModifierFunction.MOVESPEED_BONUS_PERCENTAGE,
            ModifierFunction.ATTACKSPEED_BONUS_CONSTANT,
            ModifierFunction.PREATTACK_BONUS_DAMAGE
        ];
    }

    GetModifierAttackSpeedBonus_Constant() {
        return this.asSpeed;
    }

    GetModifierMoveSpeedBonus_Percentage() {
        return this.speedBuffPct;
    }

    GetModifierPreAttack_BonusDamage() {
        return this.damage;
    }

    GetTexture() {
        return "item_yasha";
    }
}
